core/src/main/java/org/apache/iceberg/rest/responses/PlanTableScanResponse.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.rest.responses;
+
+import java.util.List;
+import java.util.Map;
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.rest.PlanStatus;
+
+public class PlanTableScanResponse extends BaseScanResponse {
+  private final PlanStatus planStatus;
+  private final String planId;
+
+  private PlanTableScanResponse(
+      PlanStatus planStatus,
+      String planId,
+      List<String> planTasks,
+      List<FileScanTask> fileScanTasks,
+      List<DeleteFile> deleteFiles,
+      Map<Integer, PartitionSpec> specsById) {
+    super(planTasks, fileScanTasks, deleteFiles, specsById);
+    this.planStatus = planStatus;
+    this.planId = planId;
+    validate();
+  }
+
+  public PlanStatus planStatus() {
+    return planStatus;
+  }
+
+  public String planId() {
+    return planId;
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("planStatus", planStatus())
+        .add("planId", planId())
+        .toString();
+  }
+
+  @Override
+  public void validate() {
+    Preconditions.checkArgument(
+        planStatus() != null, "Invalid response: plan status must be defined");
+    Preconditions.checkArgument(
+        planStatus() != PlanStatus.SUBMITTED || planId() != null,
+        "Invalid response: plan id should be defined when status is 'submitted'");
+    Preconditions.checkArgument(
+        planStatus() != PlanStatus.CANCELLED,
+        "Invalid response: 'cancelled' is not a valid status for planTableScan");
+    Preconditions.checkArgument(
+        planStatus() == PlanStatus.COMPLETED || (planTasks() == null && fileScanTasks() == null),
+        "Invalid response: tasks can only be returned in a 'completed' status");
```

Comment:
nit on some of these messages: I think including the word "returned" in the message isn't really the right way to express the failure. Yes, typically this will be because a server returned some invalid data and the client can't build the model. But more generally, error messages at this abstraction shouldn't care about the overall context. What if I'm implementing a server and I just build a bad `PlanTableScanResponse`? Nothing is really being returned etc.

So I think something like "Invalid response: tasks can only be *defined* when status is completed" is better.

So I'd probably go through and just double check some of these because we're assuming some context (albeit totally reasonable/typical context) in these error messages.